include/Basic/SourceManager.hpp

```diff
@@ -0,0 +1,236 @@
+
+namespace glu {
+
+///
+/// @class ContentCache
+/// @brief A file loaded will be cached in this class.
+///
+/// The ContentCache class is responsible for caching the content of a file that
+/// has been loaded. It is used by the SourceManager to store the content of a
+/// file. The content of a file is stored in a MemoryBuffer object which is
+/// owned by the ContentCache.
+///
+class alignas(8) ContentCache {
+
+private:
+    llvm::StringRef _fileName;
+
+    /// The buffer containing the content of the file.
+    /// Is nullptr if the file has not been loaded or if the buffer is invalid.
+    std::unique_ptr<llvm::MemoryBuffer> _buffer;
+
+public:
+    ContentCache(llvm::StringRef fileName)
+        : _fileName(fileName), _buffer(nullptr)
+    {
+    }
+
+    /// Copy constructor and assignment operator are deleted.
+    /// A unique owned file content should not be copied.
+    ContentCache(ContentCache const &other) = delete;
+    ContentCache &operator=(ContentCache const &other) = delete;
+
+    /// Get the size of the loaded buffer if loaded.
+    unsigned getSize() const
+    {
+        return _buffer ? (unsigned) _buffer->getBufferSize() : 0;
+    }
+
+    /// Return the buffer, only if it has been loaded.
+    std::optional<llvm::MemoryBufferRef> getBufferIfLoaded() const
+    {
+        if (_buffer)
+            return _buffer->getMemBufferRef();
+        return std::nullopt;
+    }
+
+    /// Return a StringRef to the source buffer data, only if it has already
+    /// been loaded.
+    std::optional<llvm::StringRef> getBufferDataIfLoaded() const
+    {
+        if (_buffer)
+            return _buffer->getBuffer();
+        return std::nullopt;
+    }
+
+    /// Set the MemoryBuffer for this ContentCache.
+    void setBuffer(std::unique_ptr<llvm::MemoryBuffer> buffer)
+    {
+        _buffer = std::move(buffer);
+    }
+
+    void setFileName(llvm::StringRef fileName) { _fileName = fileName; }
+
+    llvm::StringRef getFileName() const { return _fileName; }
+};
+
+///
+/// @class FileInfo
+/// @brief Contains concrete information about a file.
+///
+/// The FileInfo class contains information about a file that has been loaded by
+/// the SourceManager. It contains the SourceLocation of the import directive
+/// that brought in the file, and a reference to the ContentCache object that
+/// contains the content of the file.
+///
+class FileInfo {
+
+private:
+    /// The location of the import directive that brought in the file.
+    /// If the FileInfo is for the main file this location is invalid (ID == 0).
+    SourceLocation _importLoc;
+
+    /// The content cache of the file.
+    ContentCache *_contentCache;
+
+public:
+    /// Create a FileInfo object with the given SourceLocation and ContentCache.
+    FileInfo(
+        SourceLocation includeLoc, ContentCache &contentCache,
+        llvm::StringRef fileName
+    )
+        : _importLoc(includeLoc), _contentCache(nullptr)
+    {
+        _contentCache = &contentCache;
```

Comment:
why do you take a reference??
new -> dereference -> take address -> put in unique ptr
why not just use make_unique and take the unique_ptr as parameter
or better, put it directly in the struct as you don't need an indirection? (your types can all be move-only)